servers/rendering/renderer_rd/effects/luminance.cpp

```diff
@@ -0,0 +1,255 @@
+
+	if (prefer_raster_effects) {
+		Vector<String> luminance_reduce_modes;
+		luminance_reduce_modes.push_back("\n#define FIRST_PASS\n"); // LUMINANCE_REDUCE_FRAGMENT_FIRST
+		luminance_reduce_modes.push_back("\n"); // LUMINANCE_REDUCE_FRAGMENT
+		luminance_reduce_modes.push_back("\n#define FINAL_PASS\n"); // LUMINANCE_REDUCE_FRAGMENT_FINAL
+
+		luminance_reduce_raster.shader.initialize(luminance_reduce_modes);
+		luminance_reduce_raster.shader_version = luminance_reduce_raster.shader.version_create();
+
+		for (int i = 0; i < LUMINANCE_REDUCE_FRAGMENT_MAX; i++) {
+			luminance_reduce_raster.pipelines[i].setup(luminance_reduce_raster.shader.version_get_shader(luminance_reduce_raster.shader_version, i), RD::RENDER_PRIMITIVE_TRIANGLES, RD::PipelineRasterizationState(), RD::PipelineMultisampleState(), RD::PipelineDepthStencilState(), RD::PipelineColorBlendState::create_disabled(), 0);
+		}
+	} else {
+		// Initialize luminance_reduce
+		Vector<String> luminance_reduce_modes;
+		luminance_reduce_modes.push_back("\n#define READ_TEXTURE\n");
+		luminance_reduce_modes.push_back("\n");
+		luminance_reduce_modes.push_back("\n#define WRITE_LUMINANCE\n");
+
+		luminance_reduce.shader.initialize(luminance_reduce_modes);
+		luminance_reduce.shader_version = luminance_reduce.shader.version_create();
+
+		for (int i = 0; i < LUMINANCE_REDUCE_MAX; i++) {
+			luminance_reduce.pipelines[i] = RD::get_singleton()->compute_pipeline_create(luminance_reduce.shader.version_get_shader(luminance_reduce.shader_version, i));
+		}
+
+		for (int i = 0; i < LUMINANCE_REDUCE_FRAGMENT_MAX; i++) {
+			luminance_reduce_raster.pipelines[i].clear();
+		}
+	}
+}
+
+Luminance::~Luminance() {
+	if (prefer_raster_effects) {
+		luminance_reduce_raster.shader.version_free(luminance_reduce_raster.shader_version);
+	} else {
+		luminance_reduce.shader.version_free(luminance_reduce.shader_version);
+	}
+}
+
+void Luminance::LuminanceBuffers::set_prefer_raster_effects(bool p_prefer_raster_effects) {
+	prefer_raster_effects = p_prefer_raster_effects;
+}
+
+void Luminance::LuminanceBuffers::configure(RenderSceneBuffersRD *p_render_buffers) {
+	Size2i internal_size = p_render_buffers->get_internal_size();
+	int w = internal_size.x;
+	int h = internal_size.y;
+
+	while (true) {
+		w = MAX(w / 8, 1);
+		h = MAX(h / 8, 1);
+
+		RD::TextureFormat tf;
+		tf.format = RD::DATA_FORMAT_R32_SFLOAT;
+		tf.width = w;
+		tf.height = h;
+
+		bool final = w == 1 && h == 1;
+
+		if (prefer_raster_effects) {
+			tf.usage_bits = RD::TEXTURE_USAGE_COLOR_ATTACHMENT_BIT | RD::TEXTURE_USAGE_SAMPLING_BIT;
+		} else {
+			tf.usage_bits = RD::TEXTURE_USAGE_STORAGE_BIT;
+			if (final) {
+				tf.usage_bits |= RD::TEXTURE_USAGE_SAMPLING_BIT;
+			}
+		}
+
+		RID texture = RD::get_singleton()->texture_create(tf, RD::TextureView());
+		reduce.push_back(texture);
+
+		if (final) {
+			current = RD::get_singleton()->texture_create(tf, RD::TextureView());
+			break;
+		}
+	}
+}
+
+void Luminance::LuminanceBuffers::free_data() {
```

Comment:
Shouldn't this be called in the destructor? I actually don't see it being called anywhere

Reply:
It's handled in `RenderSceneRenderBuffersRD` itself. It will call `free` (and call `configure`) when the buffer size changes and it calls free when the render buffers are cleaned up.

See here: https://github.com/godotengine/godot/blob/master/servers/rendering/renderer_rd/storage_rd/render_scene_buffers_rd.cpp#L88